File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.cs
public class Solution {
    public int LongestPalindrome(string[] words) {
        int longest = 0, same = 0;
        var count = new int[26,26];    // count of each word
        foreach(var word in words){
            int c1 = word[0] - 'a', c2 = word[1] - 'a';
            if(count[c1,c2] > 0){
                longest += 4;
                count[c1,c2]--;
                if(c1 == c2) same--;  // we used up one word with same letters
                continue;
            }
            
            count[c2,c1]++;
            if(c1 == c2) same++;  // found one word with same letters
        }

        if(same > 0) longest += 2;  // the word with same letters can be put in center
        return longest;
    }
}

// Accepted - using dictionary
public class SolutionDict {
    public int LongestPalindrome(string[] words) {
        int longest = 0, same = 0;
        var count = new Dictionary<string, int>();  // count of each word
        foreach(var word in words){
            if(count.ContainsKey(word) && count[word] > 0){
                longest += 4;
                count[word]--;
                if(word[0] == word[1]) same--;  // we used up one word with same letters
                continue;
            }

            var reversed = $"{word[1]}{word[0]}";
            if(count.ContainsKey(reversed)) count[reversed]++;
            else count[reversed] = 1;
            if(word[0] == word[1]) same++;  // found one word with same letters
        }

        if(same > 0) longest += 2;  // the word with same letters can be put in center
        return longest;
    }
}

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        longest = same = 0
        count = [[0] * 26 for _ in range(26)]    # count of each word
        for word in words:
            c1, c2 = ord(word[0]) - ord('a'), ord(word[1]) - ord('a')
            if count[c1][c2] > 0:
                longest += 4
                count[c1][c2] -= 1
                if c1 == c2: same -= 1  # we used up one word with same letters
                continue
            
            count[c2][c1] += 1
            if c1 == c2: same += 1  # found word with same letters

        if same > 0: longest += 2  # the word with same letters can be put in center
        return longest

# Accepted - using dict
class SolutionDict:
    def longestPalindrome(self, words: List[str]) -> int:
        longest = same = 0
        count = defaultdict()  # count of each word
        for word in words:
            if count.get(word, 0) > 0:
                longest += 4
                count[word] -= 1
                if word[0] == word[1]: same -= 1  # we used up one word with same letters
                continue

            reversed = word[1] + word[0]
            count[reversed] = count.get(reversed, 0) + 1
            if word[0] == word[1]: same += 1  # found word with same letters

        if same > 0: longest += 2  # the word with same letters can be put in center
        return longest